backend/src/pong/pong.service.ts

```diff
@@ -0,0 +1,155 @@
+import { Injectable } from '@nestjs/common';
+import { Ball, GameState, MovementKeys, Paddle, Position } from './Game.types';
+
+@Injectable()
+export class PongService {
+	private readonly paddleWidth = 1;
+	private readonly paddleHeight = 12;
+
+	private readonly ballRadius = 1;
+
+	// ball direction and speed %
+	private dx = -0.4;
+	private dy = 0.4;
+
+	createNewGameState(): GameState {
+		const newGameState: GameState = {
+			playerOnePaddle: {
+				position: {
+					x: 0,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			playerTwoPaddle: {
+				position: {
+					x: 100,
+					y: 50,
+				},
+				width: this.paddleWidth,
+				height: this.paddleHeight,
+			},
+
+			ball: {
+				position: {
+					x: 40,
+					y: 65,
+				},
+				radius: this.ballRadius,
+			},
+			canvas: {
+				x: 1200,
+				y: 960,
+			},
+		};
+		return newGameState;
+	}
+
+	doesHitWall(radius: number, pos: number, direction: number): boolean {
+		const fullHeigthOrWidth = 100;
+		return (
+			pos + direction > fullHeigthOrWidth - radius || pos + direction < radius
+		);
+	}
+
+	doesHitPaddle(ball: Ball, rect: Paddle, canvas: Position) {
+		const paddleY = rect.position.y;
+		const paddleHeight = rect.height;
+		const paddleTop = paddleY - paddleHeight / 2;
+		const paddleBottom = paddleY + paddleHeight / 2;
+		const magicOffset = 0.2;
```

Comment:
it sounds like `magicOffset` is related to the ball radius, or is it actual magic somehow?
And if it is related to ball radius, should we use that to calculate it?

Reply:
changed it to actually make sense haha
`magicOffset => ballLineThicknessOffset`
This is for the visual aspect of the collision, it seems the thickness of the line is not part of the size of the circle.